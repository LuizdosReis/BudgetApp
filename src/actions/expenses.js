import uuid from 'uuid';

export const add = ({
  description = '', category = null, createAt = 0, amount = 0,
} = {}) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id: uuid(),
    description,
    category,
    createAt,
    amount,
  },
});

export const remove = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id,
});

export const edit = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates,
});
