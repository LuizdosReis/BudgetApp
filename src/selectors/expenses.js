import moment from 'moment';

export default (expenses, {
  startDate, endDate, description, sortBy,
}) => expenses
  .filter((expense) => {
    const createAtMoment = moment(expense.createAt);
    const startDateMatch = startDate ? createAtMoment.isSameOrAfter(startDate, 'day') : true;
    const endDateMatch = endDate ? createAtMoment.isSameOrBefore(endDate, 'day') : true;
    const descriptionMatch = description
      ? expense.description.toLowerCase().includes(description.toLowerCase())
      : true;
    return startDateMatch && endDateMatch && descriptionMatch;
  })
  .sort((a, b) => {
    if (sortBy === 'description') {
      return a.description < b.description ? -1 : 1;
    }
    if (sortBy === 'amount') {
      return a.amount < b.amount ? -1 : 1;
    }
    return a.createAt < b.createAt ? -1 : 1;
  });
