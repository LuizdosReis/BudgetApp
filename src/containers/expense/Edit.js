import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { edit, remove } from '../../actions/expenses';
import Form from './Form';
import Header from '../../pages/Header';

export class Edit extends React.Component {
  onSubmit = (expense) => {
    const { editExpense, history } = this.props;
    // eslint-disable-next-line react/destructuring-assignment
    editExpense(this.props.expense.id, expense);
    history.push('/expenses');
  };

  onRemove = () => {
    const { removeExpense, history, expense } = this.props;
    removeExpense({ id: expense.id });
    history.push('/expenses');
  };

  render() {
    const { expense } = this.props;
    return (
      <div id="layout">
        <Header title="Altere hÃ¡ despesa" />
        <main className="content">
          <Form expense={expense} onSubmit={this.onSubmit} />
          <button type="submit" onClick={this.onRemove}>
            Remover
          </button>
        </main>
      </div>
    );
  }
}

Edit.propTypes = {
  expense: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  editExpense: PropTypes.func.isRequired,
  removeExpense: PropTypes.func.isRequired,
};

const mapDispatchToProps = dispatch => ({
  editExpense: (id, expense) => dispatch(edit(id, expense)),
  removeExpense: data => dispatch(remove(data)),
});

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Edit);
