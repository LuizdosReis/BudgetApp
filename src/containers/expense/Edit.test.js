import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Edit } from './Edit';
import expenses from '../../fixtures/expenses';

let editExpenseSpy;
let removeExpenseSpy;
let historySpy;
let wrapper;

beforeEach(() => {
  editExpenseSpy = jest.fn();
  removeExpenseSpy = jest.fn();
  historySpy = { push: jest.fn() };
  // eslint-disable-next-line prefer-destructuring
  wrapper = shallow(
    <Edit
      editExpense={editExpenseSpy}
      removeExpense={removeExpenseSpy}
      history={historySpy}
      expense={expenses[2]}
    />,
  );
});

test('should render Edit correctly', () => {
  expect(toJson(wrapper)).toMatchSnapshot();
});

test('should handle submit', () => {
  wrapper.find('Form').prop('onSubmit')(expenses[2]);
  expect(editExpenseSpy).toHaveBeenLastCalledWith(expenses[2].id, expenses[2]);
  expect(historySpy.push).toHaveBeenLastCalledWith('/expenses');
});

test('should handle remove', () => {
  const { id } = expenses[2];
  wrapper.find('button').simulate('click');
  expect(removeExpenseSpy).toHaveBeenLastCalledWith({ id });
  expect(historySpy.push).toHaveBeenLastCalledWith('/expenses');
});
