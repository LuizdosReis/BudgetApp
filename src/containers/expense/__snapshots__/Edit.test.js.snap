// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Edit correctly 1`] = `
<div
  id="layout"
>
  <Header
    title="Altere hÃ¡ despesa"
  />
  <main
    className="content"
  >
    <Form
      expense={
        Object {
          "amount": 32.5,
          "category": "Bank",
          "createAt": 86400000,
          "description": "Credit car",
          "id": 3,
        }
      }
      onSubmit={[Function]}
    />
    <button
      onClick={[Function]}
      type="submit"
    >
      Remover
    </button>
  </main>
</div>
`;
